### PR TITLE
fix swagger-ui

### PR DESCRIPTION
**Pull request by kinglionsoft**
Even if AbpAuthorizeAttribute has no Permissions, the Api/Action on swagger-ui should also has Security property.

### CODE REVIEW

**Comment by ismcagdas** [comment]
@kinglionsoft I think we should add 403 to response if permissions.Any(), what do you think.

**Comment by kinglionsoft** [comment]
@ismcagdas  
[AbpAuthorize] means that the specfic Action can be called by users who have logined in, e.g [ConfigurationAppService](https://github.com/aspnetboilerplate/module-zero-core-template/blob/master/aspnet-core/src/AbpCompanyName.AbpProjectName.Application/Configuration/ConfigurationAppService.cs#L9). 
But without Security property, swagger-ui shows the API is open for all users includes anonymous users.

**Comment by ismcagdas** [comment]
@kinglionsoft I understand the reason of your PR. This is what I mean;

```
if (controllerAbpAuthorizeAttrs.Any() || actionAbpAuthorizeAttrs.Any())
{
	operation.Responses.Add("401", new Response { Description = "Unauthorized" });
	
	var permissions = controllerAbpAuthorizeAttrs.Union(actionAbpAuthorizeAttrs)
	.SelectMany(p => p.Permissions)
	.Distinct();
	
	if(permissions.Any()){
		operation.Responses.Add("403", new Response { Description = "Forbidden" });
	}
	
	.....
}	
```

If there is only AbpAuthorize attribute, we souldn't add 403 error to response.

**Comment by kinglionsoft** [comment]
Yes, I got. Thanks for for your guidance。

**Comment by ismcagdas** [comment]
@acjh is this ok for you ?

File: aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs [hunk 1]
```diff
@@ -10,22 +10,28 @@ public class SecurityRequirementsOperationFilter : IOperationFilter
...
+            var controllerAbpAuthorizeAttrs = context.ApiDescription.ControllerAttributes()
+                .OfType<AbpAuthorizeAttribute>();
 
-            if (permissions.Any())
+            if (controllerAbpAuthorizeAttrs.Any() || actionAbpAuthorizeAttrs.Any())
             {
                 operation.Responses.Add("401", new Response { Description = "Unauthorized" });
-                operation.Responses.Add("403", new Response { Description = "Forbidden" });
-
+
```
Comment (acjh) [inline aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs]: Remove whitespace (leave the newline).

File: aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs [hunk 2]
```diff
@@ -10,22 +10,28 @@ public class SecurityRequirementsOperationFilter : IOperationFilter
...
             {
                 operation.Responses.Add("401", new Response { Description = "Unauthorized" });
-                operation.Responses.Add("403", new Response { Description = "Forbidden" });
-
+                
+                var permissions = controllerAbpAuthorizeAttrs.Union(actionAbpAuthorizeAttrs)
+                    .SelectMany(p => p.Permissions)
+                    .Distinct();
+                    
+                if(permissions.Any()){
```
Comment (acjh) [inline aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs]: Spacing before `(` and newline before `{`:

```c#
if (permissions.Any())
{
```

File: aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs [hunk 3]
```diff
@@ -10,22 +10,28 @@ public class SecurityRequirementsOperationFilter : IOperationFilter
...
         {
-            var controllerPermissions = context.ApiDescription.ControllerAttributes()
-                .OfType<AbpAuthorizeAttribute>()
-                .Select(attr => attr.Permissions);
-
-            var actionPermissions = context.ApiDescription.ActionAttributes()
-                .OfType<AbpAuthorizeAttribute>()
-                .Select(attr => attr.Permissions);
+            var actionAttrs = context.ApiDescription.ActionAttributes();
+            if (actionAttrs.OfType<AbpAllowAnonymousAttribute>().Any())
```
Comment (acjh) [inline aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs]: `AbpAllowAnonymousAttribute` can be on controller too, right?

Comment (ismcagdas) [reply aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs]: @acjh yes 😄

Comment (kinglionsoft) [reply aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs]: Attributes on action have highest priority. AbpAuthorizeAttribute on action will cover AbpAllowAnonymousAttribute  on controller. Am I right?

File: aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs (hunk 2 above)
Comment (kinglionsoft) [reply aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs]: Got. Thanks.

File: aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs (hunk 3 above)
Comment (acjh) [reply aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs]: If by *cover*, you mean *take priority over*, then yes.

```c#
if (!actionAbpAuthorizeAttrs.Any() && controllerAttrs.OfType<AbpAllowAnonymousAttribute>().Any())
{
    return;
}
```

Comment (kinglionsoft) [reply aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs]: Without global Authorization filters, this condition is embodied by:
```c#
if (controllerAbpAuthorizeAttrs.Any() || actionAbpAuthorizeAttrs.Any())
{
}
```
I don't think that global Authorization filters is a good idea on ABP.
And, I will take your advise. Anyway, it does not affect the performance of the application.

File: aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs [hunk 4]
```diff
@@ -10,22 +10,34 @@ public class SecurityRequirementsOperationFilter : IOperationFilter
...
+            if (controllerAbpAuthorizeAttrs.Any() || actionAbpAuthorizeAttrs.Any())
             {
                 operation.Responses.Add("401", new Response { Description = "Unauthorized" });
-                operation.Responses.Add("403", new Response { Description = "Forbidden" });
-
+                
+                var permissions = controllerAbpAuthorizeAttrs.Union(actionAbpAuthorizeAttrs)
+                    .SelectMany(p => p.Permissions)
+                    .Distinct();
+
```
Comment (acjh) [inline aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs]: Remove whitespace (leave the newline).

File: aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs [hunk 5]
```diff
@@ -10,22 +10,34 @@ public class SecurityRequirementsOperationFilter : IOperationFilter
...
+                
+                var permissions = controllerAbpAuthorizeAttrs.Union(actionAbpAuthorizeAttrs)
+                    .SelectMany(p => p.Permissions)
+                    .Distinct();
+                    
+                if (permissions.Any())
+                {
+                    operation.Responses.Add("403", new Response { Description = "Forbidden" });
+                }
+
```
Comment (acjh) [inline aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs]: Remove whitespace (leave the newline).

File: aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs (hunk 1 above)
Comment (acjh) [reply aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Host/Startup/SecurityRequirementsOperationFilter.cs]: ^

**Comment by ismcagdas** [comment]
@kinglionsoft can you make the changes requested by @acjh when you have time ? If you don't have time, we can merge this PR and make the changes after that.

**Comment by kinglionsoft** [comment]
Is that OK?

**Comment by ismcagdas** [comment]
Thanks a lot both of you @acjh and @kinglionsoft 😄 .

**Pull request merged**